nemo/collections/tts/modules/submodules.py

```diff
@@ -410,3 +419,284 @@ def forward(self, forward_input: Tuple[torch.Tensor, torch.Tensor]):
+        self.softmax = torch.nn.Softmax(dim=2)
+        self.temperature = (self.output_size // self.n_head) ** 0.5
+        torch.nn.init.normal_(self.tokens)
+
+    @property
+    def input_types(self):
+        return {
+            "inputs": NeuralType(('B', 'D'), EncodedRepresentation()),
+            "token_id": NeuralType(('B'), Index(), optional=True),
+        }
+
+    @property
+    def output_types(self):
+        return {
+            "style_emb": NeuralType(('B', 'D'), EncodedRepresentation()),
+        }
+
+    def forward(self, inputs):
+        bs = inputs.size(0)
+        q = self.q_linear(inputs.unsqueeze(1))
+        k = self.k_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+        v = self.v_linear(self.tanh(self.tokens).unsqueeze(0).expand(bs, -1, -1))
+
+        q = q.view(bs, q.shape[1], self.n_head, self.token_size)
+        k = k.view(bs, k.shape[1], self.n_head, self.token_size)
+        v = v.view(bs, v.shape[1], self.n_head, self.token_size)
+
+        q = q.permute(2, 0, 1, 3).contiguous().view(-1, q.shape[1], q.shape[3])
+        k = k.permute(2, 0, 3, 1).contiguous().view(-1, k.shape[3], k.shape[1])
+        v = v.permute(2, 0, 1, 3).contiguous().view(-1, v.shape[1], v.shape[3])
+
+        scores = torch.bmm(q, k) / self.temperature
+        scores = self.softmax(scores)
+
+        style_emb = torch.bmm(scores, v).squeeze(1)
+        style_emb = style_emb.contiguous().view(self.n_head, bs, self.token_size)
+        style_emb = style_emb.permute(1, 0, 2).contiguous().view(bs, -1)
+
+        return style_emb
+
+
+class SVEmbeddingLinearProjectionNetwork(torch.nn.Module):
+    def __init__(self, speaker_embedding_dim, symbols_embedding_dim):
+        super(SVEmbeddingLinearProjectionNetwork, self).__init__()
+        self.speaker_proj = torch.nn.Linear(speaker_embedding_dim, symbols_embedding_dim)
+        self.act_fn = torch.nn.ReLU()
+
+    def forward(self, sv_embedding):
+        x = self.speaker_proj(sv_embedding)
+        return x.unsqueeze(1)
+
+
+class SpeakerEncoder(torch.nn.Module):
+    """
+    class SpeakerEncoder represents speakers representation. This module can combine GST (global style token)
+    based speaker embeddings, speaker embeddings from speaker verificaiton models and lookup table speaker
+    embeddings.
+    """
+
+    def __init__(self, gst_module=None, sv_projection_module=None, lookup_emb_projection_module=None):
+        """
+        Constructor.
+        Parameters:
+            - gst_module: Neural module to get GST based speaker embedding
+            - sv_projection_module: Neural module to get Speaker Verification based speaker embedding
+            - lookup_emb_projection_module: Neural module to project lookup speaker embedding
+        """
+        super(SpeakerEncoder, self).__init__()
+        self.gst_module = gst_module
+        self.sv_projection_module = sv_projection_module
+        self.lookup_emb_projection_module = lookup_emb_projection_module
```

Comment:
`lookup_emb_projection_module` looks unused with example config. Do we need it?

Reply:
You are right this is unused, we kept it as an option if we want to use it. All these three components (gst_module, sv_projection_module, lookup_emb_projection_module) can be instantiated and used or not based on what we define in the config to represent speaker.  
Once we decide on if we should include lookup table in this class or not we will make the appropriate changes.